src/openforms/js/components/admin/SessionStatus.js

```diff
@@ -0,0 +1,283 @@
+    if (raw) return component;
+    return <>&nbsp;({component})</>;
+}
+
+
+const SessionStatusWrapper = ({ showInModal=false, modalTitle, submitRowContent=null, onCloseModal, children }) => {
+    const intl = useIntl();
+
+    if (!showInModal) {
+        return (
+            <div className="session-status-wrapper session-status-wrapper--no-modal">
+                {children}
+            </div>
+        );
+    }
+
+    return (
+        <FormModal isOpen title={modalTitle} closeModal={onCloseModal} onFormSubmit={(event) => {
+            event.preventDefault();
+            onCloseModal();
+        }} extraModifiers={['small']}>
+            <div className="session-status-wrapper session-status-wrapper--modal">
+                {children}
+            </div>
+            <SubmitRow isDefault>
+                {
+                    submitRowContent ?? (
+                        <SubmitAction
+                            text={intl.formatMessage({description: 'Session status modal close button', defaultMessage: 'Back to login'})}
+                        />
+                    )
+                }
+            </SubmitRow>
+        </FormModal>
+    );
+};
+
+SessionStatusWrapper.propTypes = {
+    showInModal: PropTypes.bool,
+    modalTitle: PropTypes.node,
+    children: PropTypes.node,
+    onCloseModal: PropTypes.func.isRequired,
+    submitRowContent: PropTypes.node,
+};
+
+
+const useTriggerWarning = (numSeconds, timeToExpiryInMS) => {
+    let timeout;
+
+    const [showWarning, setShowWarning] = useState(false);
+
+    // no time available
+    if (numSeconds == null) {
+        timeout = 10 * 3600 * 1000; // 10 hours as a fallback
+    } else {
+        // re-render WARN_SESSION_TIMEOUT_FACTOR before the session expires to show a warning
+        const timeToWarningFromFactorInMS = WARN_SESSION_TIMEOUT_FACTOR * numSeconds * 1000;
+        // re-render WARN_SESSION_TIMEOUT_SECONDS before the expiry timestamp
+        const timeToWarningFromFixedOffsetInMS = timeToExpiryInMS - (WARN_SESSION_TIMEOUT_SECONDS * 1000);
+
+        // it's possible the session duration is shorter than the upfront notice we give,
+        // leading to negative values for timeToWarningFromFixedOffsetInMS
+        if (timeToWarningFromFixedOffsetInMS <= 0) {
+            timeout = timeToWarningFromFactorInMS;
+        } else {
+            // grab whichever is soonest
+            timeout = Math.min(timeToWarningFromFactorInMS, timeToWarningFromFixedOffsetInMS);
+        }
+    }
+
+    // ensure we have a positive timeout value
+    const timeToWarningInMS = Math.max(timeout, 0);
+
+    const [isReady, cancel, reset] = useTimeoutFn(
+        () => setShowWarning(true),
+        timeToWarningInMS
+    );
+    return [
+        showWarning,
+        () => {
+            setShowWarning(false);
+            reset();
+        },
+    ];
+};
+
+
+const SessionStatus = () => {
```

Comment:
I find this part a bit tricky to follow. If I understand correctly:
- `SessionStatus` looks at the global state (that is set by the `onResponseHook` in the `fetch` function) and sees in how long the session will expire.
- It sets a timeout (but I am not too clear what will run at the end of this timeout? Is it the `useTimeoutFn` in `useTriggerWarning`?)
- `useTriggerWarning` checks if warnings should be shown based on the time to expiry
- If the expiry date changes then the timeout is reset
- Based on `authFailure`, `isExpired`, `showWarning`, the right message is shown.

Reply:
> I find this part a bit tricky to follow. If I understand correctly:
> 
> * `SessionStatus` looks at the global state (that is set by the `onResponseHook` in the `fetch` function) and sees in how long the session will expire.

Correct, and it also subscribes to global state updates. If the global state changes, the component re-renders.

> * It sets a timeout (but I am not too clear what will run at the end of this timeout? Is it the `useTimeoutFn` in `useTriggerWarning`?)

The `useTimeout` hook from `react-use` triggers a component re-render after the timeout. This is needed to re-calculate the derived booleans `isExpired` etc., because a fresher value for `now` is then used while the original `date` is still the same timestamp.

> * `useTriggerWarning` checks if warnings should be shown based on the time to expiry

Correct, this essentially does the same except it runs some time before the `useTimeout` that notifies you that the session _has_ expired. The warning notifies you that it _will_ expire.

> * If the expiry date changes then the timeout is reset

Correct

> * Based on `authFailure`, `isExpired`, `showWarning`, the right message is shown.

Correct

Reply:
I see! This is clear now, thank you :slightly_smiling_face: